Lexer tuple index lexed as an exponential notation

I tried this code:

```rust
struct First {
    i: usize,
}

impl First {
    pub fn e_function(self) {}
}

pub fn main() {
    let a = (First { i: 10 }, 2);
    a.0.e_function();
}
```

I expected to see this happen: No error ([link to the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=831b0c0ecd0e450d3775ba93021038ed))

Instead, this happened: 

```
./tuplething.rs:11:7: error: unknown number suffix ‘function’
   11 |     a.0.e_function();
      |       ^
crab1: note: left denotation is null; returning null for parse_expr
./tuplething.rs:11:6: error: failed to parse path expr segment of method call expr
   11 |     a.0.e_function();
      |      ^
./tuplething.rs:11:5: error: failed to parse statement or expression in block expression
   11 |     a.0.e_function();
      |     ^
./tuplething.rs:11:7: error: unrecognised token ‘float literal’ for start of item
   11 |     a.0.e_function();
      |       ^~~~~~~~~~~~
./tuplething.rs:11:7: error: failed to parse item in crate
```

### Meta

- 95dcbb25a11

--- a/gcc/rust/lex/rust-lex.cc
+++ b/gcc/rust/lex/rust-lex.cc
@@ -2295,7 +2295,17 @@ Lexer::parse_decimal_int_or_float (location_t loc)
   length += std::get<1> (initial_decimal);
 
   // detect float literal
-  if (current_char == '.' && is_float_digit (peek_input (1).value))
+  //
+  // Note:
+  //
+  // We should not use is_float_digit () for this verification but instead
+  // directly ISDIGIT because rust does not support non digit values right after
+  // a dot.
+  // The following value is not legal in rust:
+  // let a = 3.e1;
+  // A `0` should be put between the dot and the exponent to be valid
+  // (eg. 3.0e1).
+  if (current_char == '.' && ISDIGIT (peek_input (1).value))
     {
       // float with a '.', parse another decimal into it
 

--- a/gcc/testsuite/rust/compile/issue-2514.rs
+++ b/gcc/testsuite/rust/compile/issue-2514.rs
@@ -1,0 +1,14 @@
+struct First {
+    i: usize,
+}
+
+impl First {
+    pub fn e_function(mut self) {
+        self.i += 1;
+    }
+}
+
+pub fn main() {
+    let a = (First { i: 10 }, 2);
+    a.0.e_function();
+}

